### ISSUE TITLE
Can you tag an earlier release prior to 2.0.0?

### ISSUE THREAD

**Issue by arknoll** [defuse/php-encryption#268]
Within the Drupal community there is a module which relies on this library (http://cgit.drupalcode.org/real_aes/tree/composer.json?h=8.x-1.x)

It is set to dev-master, but with the updates to 2.0.0 everyone's codebase is failing. Can we get a tag for 1.2.2 to commit https://github.com/defuse/php-encryption/commit/8f72d58522ed7a1f938d43ccffee509dcb8a1ee7

That would allow us to lock down and have the composer dependencies continue to come in appropriately.

Thank you for the help.

**Comment by paragonie-scott** [comment]
I'd recommend sticking with v1.2.1, which was the last official release of the v1.x branch. If you depended on under-development functionality, you have a tough choice to make:

* Refactor to work with v1.2.1, which is acceptable.
* Refactor to work with v2.0.0, which is optimal.

A lot of improvements landed in v2.0.0: https://paragonie.com/blog/2016/05/defuse-security-s-php-encryption-library-version-2-0-0-released

**Comment by defuse** [comment]
@arknoll: Is 8f72d58 the commit you intended? That's what's tagged as v2.0.0.

**Comment by defuse** [comment]
If you mean tag the stuff that got merged into master after `1.2.1` was tagged, but before `2.0.0` got merged into master, then I can do that.

**Comment by defuse** [comment]
(The stuff that got merged into master after `1.2.1` but before `2.0.0` got merged should be API-compatible with `1.2.1`, so you should be able to just use `1.2.1`)

**Comment by larowlan** [comment]
@arknoll you should be able to use the #ref format in your composer.json
See https://getcomposer.org/doc/04-schema.md#package-links
Specifically:
```json
{
    "require": {
        "monolog/monolog": "dev-master#2eb0c0978d290a1c45346a1955188929cb4e5db7",
        "acme/foo": "1.0.x-dev#abc123"
    }
}
```

**Comment by arknoll** [comment]
I tried 1.2.1, but it didn't seem to work. Let me try again and see what the problem was.

@larowlan Using #ref doesn't work due to https://github.com/composer/composer/issues/3705

**Comment by lookitsjonno** [comment]
I too would find this really useful. Not got chance to upgrade to 2.0.0 just yet and if I run a composer update it'll break functionality. I'm currently locked to commit 522859f0b3f35fe83be5803ede83af3f517bfd5b if that helps.

**Comment by defuse** [comment]
Closing this because 1.2.1 should work. If it doesn't, then one of two things is true:

1. 1.2.1. isn't compatible with the old release version, which is a bug (please provide evidence and open a ticket).
2. You were using a version that was never actually a released version. There are three possibilities:
    1. On the `master` branch before `1.2.1` was tagged (this might be tricky to deal with).
    2. On the `master` branch before `2.0` was merged (using `1.2.1` should work).
    3. On the branch where v2.0 was being developed (I can't help you there, intermediate versions on that branch contained security bugs, so the only safe thing to do is to decrypt with the version you have and re-encrypt with the latest version). I'm not sure why this would happen, `dev-master` shouldn't pull from some random dev branch; it was never the default branch on GitHub.

I'm trying to find out where the two revisions mentioned in the comments are from.

**Comment by defuse** [comment]
I can't even find those revisions in my local clone, but GitHub seems to know what they are. :/

**Comment by defuse** [comment]
@arknoll: Your commit looks like it was from the v2 development branch:

<img width="889" alt="screen shot 2016-09-11 at 9 36 08 am" src="https://cloud.githubusercontent.com/assets/4211072/18417630/87d907f2-7803-11e6-8fee-0ec8b80c4fc2.png">

I'm not sure how that would have happened -- that commit should have only ever existed in a PR before it got merged. Does composer `dev-master` reach inside PRs!?

@lookitsjonno: Your commit looks like it was between on master before 1.2.1 was tagged (so 1.2.1 should work, since it's compatible):

<img width="868" alt="screen shot 2016-09-11 at 9 34 39 am" src="https://cloud.githubusercontent.com/assets/4211072/18417635/a4cfb356-7803-11e6-90bd-03522a360e0f.png">
